torch/distributed/_composable/fsdp/_fsdp_param_group.py

```diff
@@ -28,9 +36,141 @@ def __init__(
         self.mesh_info = mesh_info
         self.device = device
         self._training_state = TrainingState.IDLE
+        # Group's sharded state always matches its parameters' sharded states
         self._sharded_state = ShardedState.SHARDED
+        self._init_mp_dtypes()
         self._module_fqn: Optional[str] = None  # prefixed from root module
 
+        # - Communication and communication/computation overlap
+        default_stream = torch.cuda.current_stream()
+        self.default_stream: torch.cuda.Stream = default_stream
+        self.all_gather_copy_in_stream: torch.cuda.Stream = default_stream
+        self.all_gather_stream: torch.cuda.Stream = default_stream
+        self.all_gather_state = AllGatherStateHolder()
+
+        # - CUDA events for stream synchronization
+        # Holds the all-gather output buffer, sync objects, and metadata
+        self._all_gather_result: Optional[AllGatherResult] = None
+
+    # Initialization #
+    def _init_mp_dtypes(self) -> None:
+        orig_dtypes = {fsdp_param.orig_dtype for fsdp_param in self.fsdp_params}
+        if len(orig_dtypes) != 1:
+            # This can be relaxed if we copy-out for the reduce-scatter
+            raise AssertionError(
+                f"FSDP expects uniform original parameter dtype but got {orig_dtypes}"
+            )
+        self._orig_dtype = next(iter(orig_dtypes))
+        self._param_dtype = self._orig_dtype
+
+    # Runtime #
+    def unshard(self, async_op: bool = False):
```

Comment:
Curious under what circumstance is `unshard` called multiple times within an iteration for one `TrainingState`. I'm trying to understand why we want to make the method idempotent instead of erroring out on re-entrant.

Reply:
I think you are right that we should not call `unshard()` more than once for the same `TrainingState`. I can think on this more.

However, idempotence (across `unshard` calls with different `TrainingState`s) gives us one nice thing: we can have an `async_op=True` call before an `async_op=False` call.

Suppose the user wants to _manually_ prefetch the 1st all-gather that would otherwise not be overlapped with any compute. The user can add a `fsdp_module.unshard(async_op=True)` call, e.g. before some embedding all-to-all. Then, when the `fsdp_module` runs `forward()` per normal, it will call `unshard(async_op=False)`, return early since `self._all_gather_result is not None`, and call `wait_for_unshard()`, waiting on the 1st `async_op=True` call.

This `fsdp_module.unshard()` method has not been exposed from the `FSDP` class yet in these PRs, but I do plan to do that in the future. 

Let me know if this makes sense.

Reply:
Thanks for the context. It's totally fine to leave this idempotent so long as there's a useful scenario.